kobotoolbox/kpi
# Rsgraber/task 944 permissions changes

### 🗒️ Checklist

1. [ ] run linter locally
2. [ ] update all related docs (API, README, inline, etc.), if any
3. [ ] draft PR with a title `<type>(<scope>)<!>: <title> TASK-1234`
4. [ ] tag PR: at least `frontend` or `backend` unless it's global
5. [ ] fill in the template below and delete template comments
6. [ ] review thyself: read the diff and repro the preview as written
7. [ ] open PR & confirm that CI passes
8. [ ] request reviewers, if needed
9. [ ] delete this section before merging

### 📣 Summary





### 📖 Description





### 👷 Description for instance maintainers





### 👀 Preview steps



Bug template:
1. ℹ️ have an account and a project
2. do this
3. do that
4. 🔴 [on main] notice that this isn't anywhere BUT it should be here
5. 🟢 [on PR] notice that this is here

Feature/no-change template:
1. ℹ️ have at least 2 accounts and a project. For the test plan we'll assume the second user is "user2"
2. Go to Project > Settings > Sharing
3. Toggle 'Allow submissions to this form without a username and password'
4. Go to /api/v2/audit-logs/?q=log_type:project-history AND metadata__asset_uid:<uid>&format=json
5. 🟢 You should see a new log
```
app_label:	"kpi"
model_name:	"asset"
user:	"http://localhost/api/v2/users/<user>/format=json"
user_uid:	"<user uid>"
username:	"<username>"
action:	"allow-anonymous-submissions"
metadata:	
    source:	"Firefox (Ubuntu)"
    asset_uid:	"<uid>"
    ip_address:	"172.18.0.1"
    log_subtype:	"permission"
    date_created:	"<date>"
log_type	"project-history"
```
7. Toggle it off again
8. Reload the endpoint. You should see a new log with action="disallow-anonymous-submissions"
9. Check 'Anyone can view this form'
10. Reload the endpoint. You should see a new log with action="share-form-publicly"
11. Check 'Anyone can view submissions made to this form'
12. Reload the endpoint. You should see a new log with action="share-data-publicly"
13. Uncheck 'Anyone can view submissions made to this form'
10. Reload the endpoint. You should see a new log with action="unshare-data-publicly"
11. Check 'Anyone can view this form'
12. Reload the endpoint. You should see a new log with action="unshare-form-publicly"
13. Click Add User and enter 'user2'
14. Grant 'Edit Submissions from specific users' -> "user2" (this will automatically grant a few others)
15. Reload the endpoint. You should see a new log with action="modify-user-permissions" and the following in the metadata (order of lists doesn't matter):
```
permissions:
     username: "user2"
     added: [{'code':'change_submissions', 'filters': {'_submitted_by': 'user2'}, {'code':'view_submissions', 'filters': {'_submitted_by': 'user2'},{'code':'add_submissions', 'filters': {'_submitted_by': 'user2'}, 'partial_submissions', 'view_asset', 'add_submissions'],
    removed: []
```
16. Remove the user from the project (trash can icon next to their name)
17. Reload the endpoint. Because the front end sends this as multiple DELETE requests, you'll actually see 2 new audit logs, with metadata:
```
permissions:
     username: "user2"
     added: [],
    removed: ['view_asset', 'partial_submissions']
```
```
permissions:
     username: "user2"
     added: []
     removed: ['add_submissions']
```
18. 


### 💭 Notes
